### PR TITLE
[wip] greedy-divide-ctx-document

### PR DESCRIPTION
**Pull request by andreakiro**
#21 Still need to;
- [ ] Create a valid pipe for partitioning `pipe(node: NotteNode, gamma: int) -> list[NotteNode]` for generic usage
- [ ] Add a bunch of tests to make sure we cover some of the base edge cases :)

### CODE REVIEW

File: notte/pipe/main.py [hunk 1]
```diff
@@ -46,3 +48,51 @@ def forward(
...
+                partials = [partial_tree_by_path(context.node, path) for path in _paths]
+                merged = merge_trees(partials)
+                if merged is not None:
+                    trees.append(merged)
+            else:
+                path = paths[_partition[0]]
+                partial = partial_tree_by_path(context.node, path)
+                trees.append(partial)
+
+        # _forward for each tree, with incremental listing.
```
Comment (giordano-lucas) [inline notte/pipe/main.py:86]: I think that it's better to put lines `60-86` in a function `build_context_partitions`  to increase readability / lisibility

File: notte/browser/context.py [hunk 2]
```diff
@@ -16,8 +16,15 @@ def interaction_nodes(self) -> list[InteractionNode]:
     def markdown_description(self) -> str:
         return self.format(self.node, indent_level=0)
 
-    def format(self, node: NotteNode, indent_level: int = 0) -> str:
+    def format(
+        self,
+        node: NotteNode,
+        indent_level: int = 0,  # indentation level for the current node.
```
Comment (giordano-lucas) [inline notte/browser/context.py:22]: I think these should probably be computed when the Notte graph is created intead of when it is formated in markdown. In my other PR I created a new function`NotteNode.from_a11y_node` and you can use that to build the necessary variables.

File: notte/browser/node_type.py [hunk 3]
```diff
@@ -405,10 +407,13 @@ class NotteNode:
     id: str | None
     role: NodeRole | str
     text: str
-    subtree_ids: list[str] = field(init=False, default_factory=list)
     children: list["NotteNode"] = field(default_factory=list)
     attributes_pre: NodeAttributesPre = field(default_factory=NodeAttributesPre)
     attributes_post: NotteAttributesPost | None = None
+    subtree_ids: list[str] = field(init=False, default_factory=list)
+    _subtree_chars: int = -1  # length of formatted text for this node+children
```
Comment (giordano-lucas) [inline notte/browser/node_type.py:414]: Why not use `None` for default value ?

File: notte/pipe/main.py (hunk 1 above)
Comment (andreakiro) [reply notte/pipe/main.py:86]: Yes and it will also be helpful for you in data scraping 👍🏽

File: notte/browser/context.py (hunk 2 above)
Comment (andreakiro) [reply notte/browser/context.py:22]: Hmm yes and no. For `_path` I agree, all other variables are strictly correlated with the `format` function as the number of chars depends on the way you build your resulting output string. There is no way for me to predict this ahead of time (before calling `format()` I mean)

File: notte/browser/node_type.py (hunk 3 above)
Comment (andreakiro) [reply notte/browser/node_type.py:414]: Because it is annoying to have to deal with based-pyright type checks when something can be of two different types. Here at least it's always an integer, and my condition is that that thing is positive, that's it.

File: notte/browser/context.py (hunk 2 above)
Comment (giordano-lucas) [reply notte/browser/context.py:22]: yeah that's right actually

**Comment by andreakiro** [comment]
#56 up to date
